azureeyemodule/app/util/time_aligned_buffer.cpp

```diff
@@ -0,0 +1,150 @@
+// Standard library includes
+#include <vector>
+
+// Local includes
+#include "time_aligned_buffer.hpp"
+
+namespace timebuf {
+
+
+TimeAlignedBuffer::TimeAlignedBuffer(const cv::Mat &default_item)
+    : n_timestamped_frames(10), default_value(default_item)
+{
+    // Nothing to do
+}
+
+void TimeAlignedBuffer::put(const timestamped_frame_t &frame_and_ts)
+{
+    // Add the item
+    if (this->index >= this->timestamped_frames.size())
+    {
+        // Add a new slot
+        this->timestamped_frames.push_back(frame_and_ts);
+    }
+    else
+    {
+        // Overwrite an old item's slot
+        this->timestamped_frames.at(this->index) = frame_and_ts;
+    }
+
+    this->index += 1;
+    if (this->index >= this->n_timestamped_frames)
+    {
+        this->index = 0;
+    }
+}
+
+size_t TimeAlignedBuffer::size() const
+{
+    return this->timestamped_frames.size();
+}
+
+std::vector<cv::Mat> TimeAlignedBuffer::get_best_match_and_older(int64_t timestamp)
+{
+    // If there is nothing in the buffer yet, let's return a default value.
+    if (this->timestamped_frames.size() == 0)
+    {
+        #ifdef DEBUG_TIME_ALIGNMENT
+            util::log_debug("New Inference: No frames in buffer. Sending cached frame.");
+        #endif
+        return this->default_value;
+    }
+
+    cv::Mat oldest_frame;
+    int64_t oldest_ts = LLONG_MAX;
+
+    int64_t best_match_ts = 0;
+    int64_t smallest_time_difference = LLONG_MAX;
+
+    // Sanity check
+    assert(this->timestamped_frames.size() != 0);
+
+    // Search through the circular buffer for the oldest timestamp and the best matching timestamp.
+    for (const auto &tup : this->timestamped_frames)
+    {
+        auto tup_ts = std::get<1>(tup);
+        if (tup_ts < oldest_ts)
+        {
+            std::tie(oldest_frame, oldest_ts) = tup;
+        }
+
+        // Also find the best match
+        auto timedelta = (tup_ts > timestamp) ? (tup_ts - timestamp) : (timestamp - tup_ts);
+        if (timedelta < smallest_time_difference)
+        {
+           best_match_ts = std::get<1>(tup);
+           smallest_time_difference = timedelta;
```

Comment:
The buffer is in order. We want the smallest time difference because (in my opinion), that is the best match. I.e., if I have frames A, B, C, D, E in the buffer, and I want to match an inference at some timestamp X with those the closest matching frame, I need to look at the timestamp for A, then B, then C, etc., until I find the one that is closest in time to X.

You are right that because they are guaranteed to be in order, we could probably optimize this loop by breaking from it as soon as timedelta starts growing (because as we approach the best matching frame, the timedelta will be shrinking, and then we will find the best matching frame, and then the timedelta will start growing, at which point we could break from the loop), but why bother?